src/Jupyter/client/kernel.ts

```diff
@@ -0,0 +1,137 @@
+                regex: /\W(?:let|set|w\/|new|not|and|or|using|borrowing|newtype|mutable)\b/
+            },
+            {
+                token: "meta",
+                regex: /[^\w(\s]*(?:Int|BigInt|Double|Bool|Qubit|Pauli|Result|Range|String|Unit)\b/
+            },
+            {
+                token: "atom",
+                regex: /\W(?:true|false|Pauli(I|X|Y|Z)|One|Zero)\b/
+            },
+            {
+                token: "builtin",
+                regex: /(\\n|\W)(?:X|Y|Z|H|HY|S|T|SWAP|CNOT|CCNOT|MultiX|R|RFrac|Rx|Ry|Rz|R1|R1Frac|Exp|ExpFrac|Measure|M|MultiM)\b/
+            },
+            {
+                token: "builtin",
+                regex: /(\\n|\W)(?:Message|Length|Assert|AssertProb|AssertEqual)\b/
+            }
+        ]
+    });
+    codeMirror.defineMIME("text/x-qsharp", "qsharp");
+}
+
+class Kernel {
+    hostingEnvironment : string | undefined;
+    iqsharpVersion : string | undefined;
+
+    constructor() {
+        IPython.notebook.kernel.events.on("kernel_ready.Kernel", args => {
+            this.requestEcho();
+            this.requestClientInfo();
+        });
+    }
+
+    requestEcho() {
+        // Try sending something for the kernel to echo back in order to test
+        // communicates with the kernel. Note that iqsharp_echo_request will get
+        // two responses: an output over iopub, and a reply over shell. We thus
+        // subscribe both callbacks in order to make sure that both work
+        // correctly.
+        let value = "hello!";
+        // The output callback is registered with the kernel object itself,
+        // since outputs aren't a reply to any particular message.
+        IPython.notebook.kernel.register_iopub_handler(
+            "iqsharp_echo_output",
+            message => {
+                console.log("Got echo output:", message);
+            }
+        );
+        // By contrast, callbacks for replies are associated with the message
+        // itself, since we don't want the callback to pick up echo replies that
+        // are replies to other messages.
+        IPython.notebook.kernel.send_shell_message(
+            "iqsharp_echo_request",
+            {value: value},
+            {
+                shell: {
+                    reply: (message) => {
+                        console.log("Got echo reply:", message);
+                    }
+                }
+            }
+        );
+    }
+
+    requestClientInfo() {
+        // The other thing we will want to do as the kernel starts up is to
+        // pass along information from the client that would not otherwise be
+        // available. For example, the browser user agent isn't exposed to the
+        // kernel by the Jupyter protocol, since the client may not even be in a
+        // browser at all.
+        IPython.notebook.kernel.send_shell_message(
+            "iqsharp_clientinfo_request",
+            {
+                "user_agent": navigator.userAgent
+            },
```

Comment:
similar comment as in ClientInfoHandler, maybe we should stick to the camelCase naming convention for the JSON messages ("userAgent" instead of "user_agent") but it should be ok for the message type to be "iqsharp_clientinfo_request" as it's a literal.

Reply:
Resolving as per above conversation.